src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java

```diff
@@ -13,12 +34,368 @@
+
+    /**
+     * List of itemStacks the deliveryman will deliver to the requesting building.
+     */
+    private List<ItemStack> itemsToDeliver = new ArrayList<>();
+
+    /**
+     * Next target the deliveryman should gather stuff at.
+     */
+    private BlockPos gatherTarget = null;
+
+    /**
      * Initialize the deliveryman and add all his tasks.
      *
      * @param deliveryman the job he has.
      */
     public EntityAIWorkDeliveryman(@NotNull final JobDeliveryman deliveryman)
     {
         super(deliveryman);
+        super.registerTargets(
+                /**
+                 * Check if tasks should be executed.
+                 */
+                new AITarget(this::checkIfExecute, IDLE),
+                new AITarget(IDLE, () -> START_WORKING),
+                new AITarget(START_WORKING, this::checkWareHouse),
+                new AITarget(PREPARE_DELIVERY, this::prepareDelivery),
+                new AITarget(GATHER_IN_WAREHOUSE, this::gatherItemsFromWareHouse),
+                new AITarget(DELIVERY, this::deliver),
+                new AITarget(GATHERING, this::gather),
+                new AITarget(DUMPING, this::dump)
+
+        );
+        worker.setSkillModifier(2 * worker.getCitizenData().getEndurance() + worker.getCitizenData().getCharisma());
+        worker.setCanPickUpLoot(true);
+    }
+
+    /**
+     * Gather items from a random hut which the hut doesn't need.
+     * @return the next state to go to.
+     */
+    public AIState gather()
+    {
+        if(gatherTarget == null)
+        {
+            gatherTarget = getRandomBuilding();
+        }
+
+        if(gatherTarget == null)
+        {
+            return GATHERING;
+        }
+
+        if(!worker.isWorkerAtSiteWithMove(gatherTarget, MIN_DISTANCE_TO_WAREHOUSE))
+        {
+            return GATHERING;
+        }
+
+        Colony colony = getOwnBuilding().getColony();
+        if(colony != null)
+        {
+            AbstractBuilding building = colony.getBuilding(gatherTarget);
+            gatherFromBuilding(building);
+            return DUMPING;
+        }
+        return START_WORKING;
+    }
+
+    /**
+     * Gather not needed Items from building.
+     * @param building building to gather it from.
+     */
+    private void gatherFromBuilding(final AbstractBuilding building)
+    {
+        for(int i = 0 ; i < building.getTileEntity().getSizeInventory(); i++)
+        {
+            ItemStack stack = building.getTileEntity().getStackInSlot(i);
+
+            if(stack == null)
+            {
+                continue;
+            }
+
+            boolean needsItem = false;
+            if(building instanceof AbstractBuildingWorker && ((AbstractBuildingWorker) building).neededForWorker(stack))
+            {
+                needsItem = true;
+            }
+            //todo count how much of the itemStack has to be left behind.
+            for(Map.Entry<ItemStorage, Integer> entry: building.needXForWorker().entrySet())
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'entry' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)